File: 2022/day_08/puzzle.js
const { data_test } = require('./data_test.js');
const { puzzle_data } = require('./puzzle_data.js');

function compute_data(data) {
    let map = []
    let lines = data.split("\n")
    for (let i=0; i<lines.length; i++) {
        let line = lines[i]
        let row = []
        for (let j=0; j<line.length; j++) {
            row.push(line[j])
        }
        map.push(row)
    }
    return map
}

function is_visible(x, y, map) {
    let upOk = true;
    let downOk = true;
    let leftOk = true;
    let rightOk = true;

    for (let i=0; i < x; i++) {
        if (map[i][y] >= map[x][y]) {
            upOk = false
            break
        }
    }
    
    for (let i=map.length-1; i > x; i--) {
        if (map[i][y] >= map[x][y]) {
            downOk = false
            break
        }
    }
    
    for (let i=0; i < y; i++) {
        if (map[x][i] >= map[x][y]) {
            rightOk = false
            break
        }
    }
    
    for (let i=map[x].length-1; i > y; i--) {
        if (map[x][i] >= map[x][y]) {
            leftOk = false
            break
        }
    }

    return upOk || downOk || leftOk || rightOk
}

function get_scenic_score(x, y, map) {
    let upOk = 0;
    let downOk = 0;
    let leftOk = 0;
    let rightOk = 0;

    for (let i=x+1; i < map.length; i++) {
        downOk++
        if (map[i][y] >= map[x][y]) {
            break
        }
    }
    
    for (let i=x-1; i >= 0; i--) {
        upOk++
        if (map[i][y] >= map[x][y]) {
            break
        }
    }
    
    for (let i=y+1; i < map[x].length; i++) {
        leftOk++
        if (map[x][i] >= map[x][y]) {
            break
        }
    }
    
    for (let i=y-1; i >= 0; i--) {
        rightOk++
        if (map[x][i] >= map[x][y]) {
            break
        }
    }

    return upOk * downOk * leftOk * rightOk
}

function part1(data) {
    let res = (data.length-1) * 4
    for (let x=1; x<data.length-1; x++) {
        for (let y=1; y<data[x].length-1; y++) {
            if (is_visible(x, y, data)) {
                res += 1
            }
        }
    }
    return res
}

function part2(data) {
    let res = 0
    for (let x=0; x<data.length; x++) {
        for (let y=0; y<data[x].length; y++) {
            res = Math.max(res, get_scenic_score(x, y, data))
        }
    }
    return res
}


let data = compute_data(puzzle_data)
console.log('Result for puzzle 1 = ', part1(data))
console.log('Result for puzzle 2 = ', part2(data))

File: 2024/day_08/puzzle.py
def read_data(filename):
    return open(f"{filename}").read().strip()

###############################
## Construct data for solver ##
###############################
def constructData(data):
    m = {}

    lines = data.split("\n")
    maxX = len(lines)
    for x in range(len(lines)):
        row = lines[x]
        maxY = len(row)
        for y in range(len(list(row))):
            c = list(row)[y]
            if c != ".":
                old = m[c] if c in m else []
                m[c] = old
                m[c].append([x,y])
    return m, maxX, maxY


###########
## UTILS ##
###########
def find_antinode(x1, y1, x2, y2, MAX):
    distX = x2 - x1
    distY = y2 - y1
    a1 = [x1-distX, y1-distY]
    a2 = [x2+distX, y2+distY]

    ok = []
    for a in [a1, a2]:
        if in_grid(a[0], a[1], MAX[0], MAX[1]):
            ok.append(a)
    return ok

def find_all_antinode(x1, y1, x2, y2, MAX):
    distX = x2 - x1
    distY = y2 - y1
    a1 = [x1-distX, y1-distY]
    a2 = [x2+distX, y2+distY]

    ok = [[x1, y1], [x2, y2]]

    while in_grid(a1[0], a1[1], MAX[0], MAX[1]):
        ok.append(a1)
        a1 = [a1[0]-distX, a1[1]-distY]

    while in_grid(a2[0], a2[1], MAX[0], MAX[1]):
        ok.append(a2)
        a2 = [a2[0]+distX, a2[1]+distY]

    return ok

def in_grid(x, y, MAX_X, MAX_Y):
    return (0 <= x < MAX_X) and (0 <= y < MAX_Y)


###################
## Solver part 1 ##
###################
def part1(data):
    data, MAX_X, MAX_Y = data
    res = []
    for key in data:
        for i in range(len(data[key])-1):
            for j in range(i+1, len(data[key])):
                p1 = data[key][i]
                p2 = data[key][j]
                antinodes = find_antinode(p1[0], p1[1], p2[0], p2[1], [MAX_X, MAX_Y])
                for a in antinodes:
                    if not str(a) in res:
                        res.append(str(a))
    return len(res)


###################
## Solver part 2 ##
###################
def part2(data):
    data, MAX_X, MAX_Y = data
    res = []
    for key in data:
        for i in range(len(data[key])-1):
            for j in range(i+1, len(data[key])):
                p1 = data[key][i]
                p2 = data[key][j]
                antinodes = find_all_antinode(p1[0], p1[1], p2[0], p2[1], [MAX_X, MAX_Y])
                for a in antinodes:
                    if not str(a) in res:
                        res.append(str(a))
    return len(res)


##########
## MAIN ##
##########
data = read_data("puzzle_data.txt")
print('Result for puzzle 1 = ', part1(constructData(data)))
print('Result for puzzle 2 = ', part2(constructData(data)))